staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto

```diff
@@ -1531,6 +1537,8 @@ message PullImageResponse {
     // Reference to the image in use. For most runtimes, this should be an
     // image ID or digest.
     string image_ref = 1;
+    // Mountpoint is an absolute path where the OCI object is mounted on disk.
+    string mountpoint = 2;
```

Comment:
@mikebrow from our discussion last Friday: Is it correct that you'd like to see a dedicated `message` type here, which acts as typed enum to return either:

- A local path on disk (kubelet will use that path on container creation to reference a mount)
- An image ID or digest

I'm wondering how the kubelet should handle the "image ID or digest" case, do you have any details about that?

Reply:
Yes, proposal is to either add a volume / source mount type to the pull image response or a new struct that has the type. The source of an image is currently the image_ref, which may be an ID or digest, both work and we understand the usefulness of absolute path for the crio case, but that type of source reference does not fit right for containerd snapshotters. 

So at min, the proposal is to return a type value here in pull image response to indicate to the kubelet which ref should be used for the volume mount source. 

The type returned and content should then be passed in as the mount source on create container.

The mount sourceType enum should at least include imageID, imageDigest, and hostPath.

Reply:
For example:
```
CRI image volume changes

// A VolumeMountSource describes the supported source type for a volume mount.
type VolumeMountSource int32

const (
    // HostPath is the default source mount type for a volume.
    VolumeMountSource_HostPath VolumeMountSource (if supported by container runtime)= 0
    // ImageRef (image id or image digest) will be used by the runtime as the source.
    VolumeMountSource_ImageRef VolumeMountSource = 1
)

var VolumeMountSource_name = map[int32]string{
    0: "HOSTPATH",
    1: "IMAGEREF",
}

var VolumeMountSource_value = map[string]int32{
    "HOSTPATH":       0,
    "IMAGEREF": 1,
}


type Image struct {
    // ID of the image.
    Id string `protobuf:"bytes,1,opt,name=id,proto3" json:"id,omitempty"`
    // Other names by which this image is known.
    RepoTags []string `protobuf:"bytes,2,rep,name=repo_tags,json=repoTags,proto3" json:"repo_tags,omitempty"`
    // Digests by which this image is known.
    RepoDigests []string `protobuf:"bytes,3,rep,name=repo_digests,json=repoDigests,proto3" json:"repo_digests,omitempty"`
    // Size of the image in bytes. Must be > 0.
    Size_ uint64 `protobuf:"varint,4,opt,name=size,proto3" json:"size,omitempty"`
    // UID that will run the command(s). This is used as a default if no user is
    // specified when creating the container. UID and the following user name
    // are mutually exclusive.
    Uid *Int64Value `protobuf:"bytes,5,opt,name=uid,proto3" json:"uid,omitempty"`
    // User name that will run the command(s). This is used if UID is not set
    // and no user is specified when creating container.
    Username string `protobuf:"bytes,6,opt,name=username,proto3" json:"username,omitempty"`
    // ImageSpec for image which includes annotations
    Spec *ImageSpec `protobuf:"bytes,7,opt,name=spec,proto3" json:"spec,omitempty"`
    // Recommendation on whether this image should be exempt from garbage collection.
    // It must only be treated as a recommendation -- the client can still request that the image be deleted,
    // and the runtime must oblige.
    Pinned               bool     `protobuf:"varint,8,opt,name=pinned,proto3" json:"pinned,omitempty"`
 +    // SourceMountType is the container runtime supported source type for volume mounting the pulled image contents.
+    SourceMountType VolumeMountSource `protobuf:"varint,1,opt,name=mode,proto3,enum=runtime.v1.VolumeMountSource" json:"mode,omitempty"`
+    // Mountpoint is an absolute path where the OCI object is mounted on disk (if supported by container runtime).
+    string mountpoint = 2;

    XXX_NoUnkeyedLiteral struct{} `json:"-"`
    XXX_sizecache        int32    `json:"-"`
}

type PullImageResponse struct {
    // Reference to the image in use. For most runtimes, this should be an
    // image ID or digest.
    ImageRef             string   `protobuf:"bytes,1,opt,name=image_ref,json=imageRef,proto3" json:"image_ref,omitempty"`

+    // SourceMountType is the container runtime supported source type for volume mounting the pulled image contents.
+    SourceMountType VolumeMountSource `protobuf:"varint,1,opt,name=mode,proto3,enum=runtime.v1.VolumeMountSource" json:"mode,omitempty"`
+    // Mountpoint is an absolute path where the OCI object is mounted on disk (if supported by container runtime).
+    string mountpoint = 2;

    XXX_NoUnkeyedLiteral struct{} `json:"-"`
    XXX_sizecache        int32    `json:"-"`
}


// ImageSpec is an internal representation of an image.
type ImageSpec struct {
    // Container's Image field (e.g. imageID or imageDigest).
    Image string `protobuf:"bytes,1,opt,name=image,proto3" json:"image,omitempty"`
    // Unstructured key-value map holding arbitrary metadata.
    // ImageSpec Annotations can be used to help the runtime target specific
    // images in multi-arch images.
    Annotations map[string]string `protobuf:"bytes,2,rep,name=annotations,proto3" json:"annotations,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
    // The container image reference specified by the user (e.g. image[:tag] or digest).
    // Only set if available within the RPC context.
    UserSpecifiedImage string `protobuf:"bytes,18,opt,name=user_specified_image,json=userSpecifiedImage,proto3" json:"user_specified_image,omitempty"`
    // Runtime handler to use for pulling the image.
    // If the runtime handler is unknown, the request should be rejected.
    // An empty string would select the default runtime handler.
    RuntimeHandler       string   `protobuf:"bytes,19,opt,name=runtime_handler,json=runtimeHandler,proto3" json:"runtime_handler,omitempty"`

+    // MountLabel is the SELinux label to be used when using this image in a volume mount operation.
+    string mount_label = 21;

    XXX_NoUnkeyedLiteral struct{} `json:"-"`
    XXX_sizecache        int32    `json:"-"`
}
```

Reply:
ah yes and the mount change:

```
// Mount specifies a host volume to mount into a container.
type Mount struct {
    // Path of the mount within the container.
    ContainerPath string `protobuf:"bytes,1,opt,name=container_path,json=containerPath,proto3" json:"container_path,omitempty"`
    // Path of the mount on the host. If the hostPath doesn't exist, then runtimes
    // should report error. If the hostpath is a symbolic link, runtimes should
    // follow the symlink and mount the real destination to container.
    HostPath string `protobuf:"bytes,2,opt,name=host_path,json=hostPath,proto3" json:"host_path,omitempty"`
    // If set, the mount is read-only.
    Readonly bool `protobuf:"varint,3,opt,name=readonly,proto3" json:"readonly,omitempty"`
    // If set, the mount needs SELinux relabeling.
    SelinuxRelabel bool `protobuf:"varint,4,opt,name=selinux_relabel,json=selinuxRelabel,proto3" json:"selinux_relabel,omitempty"`
    // Requested propagation mode.
    Propagation MountPropagation `protobuf:"varint,5,opt,name=propagation,proto3,enum=runtime.v1.MountPropagation" json:"propagation,omitempty"`
    // UidMappings specifies the runtime UID mappings for the mount.
    UidMappings []*IDMapping `protobuf:"bytes,6,rep,name=uidMappings,proto3" json:"uidMappings,omitempty"`
    // GidMappings specifies the runtime GID mappings for the mount.
    GidMappings []*IDMapping `protobuf:"bytes,7,rep,name=gidMappings,proto3" json:"gidMappings,omitempty"`
    // If set to true, the mount is made recursive read-only.
    // In this CRI API, recursive_read_only is a plain true/false boolean, although its equivalent
    // in the Kubernetes core API is a quaternary that can be nil, "Enabled", "IfPossible", or "Disabled".
    // kubelet translates that quaternary value in the core API into a boolean in this CRI API.
    // Remarks:
    // - nil is just treated as false
    // - when set to true, readonly must be explicitly set to true, and propagation must be PRIVATE (0).
    // - (readonly == false && recursive_read_only == false) does not make the mount read-only.
    RecursiveReadOnly    bool     `protobuf:"varint,8,opt,name=recursive_read_only,json=recursiveReadOnly,proto3" json:"recursive_read_only,omitempty"`

+    // SourceMountType is the container runtime supported source type for volume mounting pulled image contents.
+    SourceMountType VolumeMountSource `protobuf:"varint,1,opt,name=mode,proto3,enum=runtime.v1.VolumeMountSource" json:"mode,omitempty"`
+    // Reference to the source image . For most runtimes, this should be an
+    // image ID or digest.
+    ImageRef             string   `protobuf:"bytes,1,opt,name=image_ref,json=imageRef,proto3" json:"image_ref,omitempty"`

    XXX_NoUnkeyedLiteral struct{} `json:"-"`
    XXX_sizecache        int32    `json:"-"`
}
```

Reply:
Thank you! Changed as suggested with some minor adjustments.